<center><h1>Exploring Toronto Neighborhoods to open an Indian Restaurant.</h1></center>

<h2>Introduction</h2>

As a part of IBM Data Science professional certification Capstone Project, we have to work on the real data to get an exprencies of what a data scientists go through in real life. I would like to explore the city Toronto and the objectives of this assignments is to define a business problem, find data in the web and, use Foursquare location data to compare different neighborhoods of Toronto to figure out which neighborhood is suitable for starting a new restaurant business. I go through all the process in a step by step manner from Business problem, data preparation to final analysis and finally will provide a conclusion that can be leveraged by the business stakeholders to make their decisions.




<h2>1. Description of the Problem</h2>

<h3>Prospects of a opening an Indian Restaurant in Toronto, Canada.</h3>

Toronto, the capital of the province of Ontario, is the most widespread Canadian city. Its diversity is reflected in Toronto’s ethnic neighborhoods such as Chinatown, Corso Italia, Greektown, Kensington Market, Koreatown, Little India, Little Italy, Little Jamaica, Little Portugal & Roncesvalles. One of the most immigrant-friendly cities in Canada with more than half of the entire Indian Canadian population be located in Toronto. it might be one of the best places to start an Indian restaurant. In this project we will go through step by step process to make a decision whether it is a good idea to open an Indian restaurant. We analyse the neighborhoods in Toronto to identify the most profitable area since the success of the restaurant depends on the people and ambience. Since we already know that Toronto shelter a greater number of Indians than any other city in Canada, it is a good idea to start the restaurant here. Let us analyse and explore whether it is a profitable idea or not. If profitable in which place we can start the restaurant, so that it gets more profit to the owner.


<h3>Who will be more interested in this project?</h3>

<h3> What type of clients or a group of people would be benefitted?</h3>
1. Business personnel who wants to invest or open an Indian restaurant in Toronto. This analysis will be a comprehensive guide to start or expand restaurants targeting the Indian crowd. <br>
2. Freelancer who loves to have their own restaurant as a side business. This analysis will give an idea, how beneficial it is to open a restaurant and what are the pros and cons of this business. <br> 
3.	Indian crowd who wants to find neighborhoods with lots of option for Indian restaurants. <br> 
4.	Business Analyst or Data Scientists, who wish to analyze the neighborhoods of Toronto using Exploratory Data Analysis and other statistical & machine learning techniques to obtain all the necessary data, perform some operations on it and, finally be able to tell a story out of it.


<h2>2. Data acquisition and cleaning</h2>


<h3>2.1 Data Sources</h3>

a) I’m using “List of Postal code of Canada: M” (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) wiki page to get all the information about the neighborhoods present in Toronto. This page has the postal code, borough & the name of all the neighborhoods present in Toronto. 

b) Then I’m using “https://cocl.us/Geospatial_data” csv file to get all the geographical coordinates of the neighborhoods. 

c) To get information about the distribution of population by their ethnicity I’m using “Demographics of Toronto” (https://en.m.wikipedia.org/wiki/Demographics_of_Toronto#Ethnic_diversity) wiki page. Using this page I’m going to identify the neighborhoods which are densely populated with Indians as it might be helpful in identifying the suitable neighborhood to open a new Indian restaurant. 

d) To get location and other information about various venues in Toronto I’m using Foursquare’s explore API. Using the Foursquare’s explore API (which gives venues recommendations), I’m fetching details about the venues up present in Toronto and collected their names, categories and locations (latitude and longitude). 
From Foursquare API (https://developer.foursquare.com/docs), I retrieved the following for each venue:
<li>Name: The name of the venue. </li>
<li>Category: The category type as defined by the API.</li>
<li>Latitude: The latitude value of the venue.</li>
<li>Longitude: The longitude value of the venue.</li>

<h3>2.2 Data Cleaning</h3>

<h4>a) Scraping Toronto Neighborhoods Table from Wikipedia</h4>

#### Scraped the following Wikipedia page, “List of Postal code of Canada: M” in order to obtain the data about the Toronto & the Neighborhoods in it.

<u> Assumptions made to attain the below DataFrame:</u>
<li>Dataframe will consist of three columns: PostalCode, Borough, and Neighborhood</li>
<li>Only the cells that have an assigned borough will be processed. Borough that is not assigned are ignored.</li>
<li>More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.</li>
<li>If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.</li>

wikipedia - package is used to scrape the data from wiki.

In [1]:
#!conda install -c conda-forge wikipedia --yes # Uncomment this if it is not already installed
import pandas as pd
import numpy as np
import wikipedia as wp
import lxml

In [2]:
html = wp.page("List of postal codes of Canada: M").html().encode("UTF-8")
CanadaPOdf = pd.read_html(html, header = 0)[0]
CanadaPOdf.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

<h4>b) Adding geographical coordinates to the neighborhoods</h4>

Next important step is adding the geographical coordinates to these neighborhoods. To do so I'm extracting the data present in the Geospatial Data csv file and I'm combining it with the existing neighborhood dataframe by merging them both based on the postal code.  

In [3]:
import io
import requests

url = "https://cocl.us/Geospatial_data"
lat_long = requests.get(url).text
lat_long_df=pd.read_csv(io.StringIO(lat_long))
lat_long_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [4]:
#Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
CanadaPOdf = CanadaPOdf[CanadaPOdf.Borough != 'Not assigned']
CanadaPOdf = CanadaPOdf.rename(columns={'Postal Code': 'Postalcode'})

#If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
#So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
for index, row in CanadaPOdf.iterrows():
    if row['Neighborhood'] == 'Not assigned':
        row['Neighborhood'] = row['Borough']


CanadaPOdf.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [5]:
CanadaPOdf = CanadaPOdf.groupby(['Borough', 'Postalcode'])['Neighborhood'].apply(list).apply(lambda x:', '.join(x)).to_frame().reset_index()
CanadaPOdf.head()

Borough Postalcode                       Neighborhood
0  Central Toronto        M4N                      Lawrence Park
1  Central Toronto        M4P                   Davisville North
2  Central Toronto        M4R  North Toronto West, Lawrence Park
3  Central Toronto        M4S                         Davisville
4  Central Toronto        M4T        Moore Park, Summerhill East

I'm renaming the columns to match the existing dataframe  

In [6]:
lat_long_df = lat_long_df.rename(columns={'Postal Code': 'Postalcode'})
lat_long_df.head()

Postalcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

After that I'm merging both the dataframe into one by matching on the postal code. 

In [7]:
toronto_DF = pd.merge(CanadaPOdf,lat_long_df, on='Postalcode')
toronto_DF = toronto_DF.rename(columns={'Neighborhood':'Neighborhood'})
toronto_DF.head()

Borough Postalcode                       Neighborhood   Latitude  \
0  Central Toronto        M4N                      Lawrence Park  43.728020   
1  Central Toronto        M4P                   Davisville North  43.712751   
2  Central Toronto        M4R  North Toronto West, Lawrence Park  43.715383   
3  Central Toronto        M4S                         Davisville  43.704324   
4  Central Toronto        M4T        Moore Park, Summerhill East  43.689574   

   Longitude  
0 -79.388790  
1 -79.390197  
2 -79.405678  
3 -79.388790  
4 -79.383160

In [8]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_DF['Borough'].unique()),
        toronto_DF.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


<h4>c) Scrap the distribution of population from Wikipedia</h4>
Another factor that can help us in deciding which neighborhood would be best option to open a restaurant is, the distribution of population based on the ethnic diversity for each neighborhood. As this helps us in identifying the neighborhoods which are densely populated with Indian crowd since that neighborhood would be an ideal place to open an Indian restaurant. 

Scraped the following Wikipedia page, “Demographics of Toronto” in order to obtain the data about the Toronto & the Neighborhoods in it. Compared to all the neighborhoods in Toronto below given neighborhoods only had considerable amount of Indian crowd. We are examing those neighborhood's population to identify the densely populated neighborhoods with Indian population.

In [9]:
html = wp.page("Demographics of Toronto").html().encode("UTF-8")


In [10]:
#TORONTO & EAST YORK population distribution by ethnicity 
TEY_population_df = pd.read_html(html, header = 0)[13]
TEY_population_df = TEY_population_df.rename(columns={'%':'Ethnic Origin 1 in %', 
                                                      '%.1':'Ethnic Origin 2 in %',
                                                     '%.2':'Ethnic Origin 3 in %',
                                                     '%.3':'Ethnic Origin 4 in %',
                                                     '%.4':'Ethnic Origin 5 in %',
                                                     '%.5':'Ethnic Origin 6 in %',
                                                     '%.6':'Ethnic Origin 7 in %',
                                                     '%.7':'Ethnic Origin 8 in %',
                                                     '%.8':'Ethnic Origin 9 in %'})
TEY_population_df

Riding  Population Ethnic Origin #1  \
0                Willowdale      117405          Chinese   
1         Eglinton-Lawrence      112925         Canadian   
2          Don Valley North      109060          Chinese   
3  Humber River-Black Creek      107725          Italian   
4               York Centre      103760         Filipino   
5           Don Valley West      101790          English   
6           Don Valley East       93170      East Indian   

   Ethnic Origin 1 in % Ethnic Origin #2  Ethnic Origin 2 in %  \
0                  25.9          Iranian                  12.1   
1                  14.7          English                  12.6   
2                  32.4      East Indian                   7.3   
3                  12.8      East Indian                   9.2   
4                  17.0          Italian                  13.4   
5                  19.2         Canadian                  15.1   
6                  10.6         Canadian                  10.4   

  Ethnic Origin #3  Ethnic Origin 3 in % Ethnic Origin #4  \
0           Korean                  10.6              NaN   
1           Polish                  12.0         Filipino   
2          Iranian                   7.3              NaN   
3         Jamaican                   8.5       Vietnamese   
4          Russian                   9.5         Canadian   
5         Scottish                  14.9            Irish   
6          English                  10.1          Chinese   

   Ethnic Origin 4 in % Ethnic Origin #5  Ethnic Origin 5 in %  \
0                   NaN              NaN                   NaN   
1                  11.0         Scottish                   9.7   
2                   NaN              NaN                   NaN   
3                   8.0         Canadian                   7.4   
4                   8.6              NaN                   NaN   
5                  14.2          Chinese                  11.2   
6                   8.9            Irish                   8.1   

  Ethnic Origin #6  Ethnic Origin 6 in % Ethnic Origin #7  \
0              NaN                   NaN              NaN   
1          Italian                   9.5            Irish   
2              NaN                   NaN              NaN   
3              NaN                   NaN              NaN   
4              NaN                   NaN              NaN   
5              NaN                   NaN              NaN   
6         Scottish                   8.0         Filipino   

   Ethnic Origin 7 in % Ethnic Origin #8  Ethnic Origin 8 in %  
0                   NaN              NaN                   NaN  
1                   9.2          Russian                   8.4  
2                   NaN              NaN                   NaN  
3                   NaN              NaN                   NaN  
4                   NaN              NaN                   NaN  
5                   NaN              NaN                   NaN  
6                   7.8              NaN                   NaN

In [11]:
#NORTH YORK population distribution by ethnicity 
North_population_df = pd.read_html(html, header = 0)[14]
North_population_df = North_population_df.rename(columns={'%':'Ethnic Origin 1 in %', 
                                                      '%.1':'Ethnic Origin 2 in %',
                                                     '%.2':'Ethnic Origin 3 in %',
                                                     '%.3':'Ethnic Origin 4 in %',
                                                     '%.4':'Ethnic Origin 5 in %',
                                                     '%.5':'Ethnic Origin 6 in %',
                                                     '%.6':'Ethnic Origin 7 in %',
                                                     '%.7':'Ethnic Origin 8 in %'})
North_population_df

Riding  Population Ethnic Origin #1  Ethnic Origin 1 in %  \
0      Scarborough Centre      110450         Filipino                  13.1   
1   Scarborough Southwest      108295         Canadian                  16.2   
2   Scarborough-Agincourt      104225          Chinese                  47.0   
3  Scarborough-Rouge Park      101445      East Indian                  16.7   
4   Scarborough-Guildwood      101115      East Indian                  18.0   
5       Scarborough North       97610          Chinese                  46.6   

  Ethnic Origin #2  Ethnic Origin 2 in % Ethnic Origin #3  \
0      East Indian                  12.2         Canadian   
1          English                  14.3            Irish   
2      East Indian                   7.4              NaN   
3         Canadian                  11.8       Sri Lankan   
4         Canadian                  11.6          English   
5      East Indian                  11.8       Sri Lankan   

   Ethnic Origin 3 in % Ethnic Origin #4  Ethnic Origin 4 in %  \
0                  11.2          Chinese                  10.7   
1                  11.5         Scottish                  10.9   
2                   NaN              NaN                   NaN   
3                  11.1          English                   9.8   
4                   9.7         Filipino                   8.5   
5                   9.4              NaN                   NaN   

  Ethnic Origin #5  Ethnic Origin 5 in % Ethnic Origin #6  \
0          English                   7.8       Sri Lankan   
1         Filipino                   9.5      East Indian   
2              NaN                   NaN              NaN   
3         Filipino                   9.3         Jamaican   
4       Sri Lankan                   7.8          Chinese   
5              NaN                   NaN              NaN   

   Ethnic Origin 6 in % Ethnic Origin #7  Ethnic Origin 7 in %  \
0                   7.0              NaN                   NaN   
1                   8.2          Chinese                   7.2   
2                   NaN              NaN                   NaN   
3                   8.4         Scottish                   7.2   
4                   7.1         Scottish                   7.0   
5                   NaN              NaN                   NaN   

  Ethnic Origin #8  Ethnic Origin 8 in %  
0              NaN                   NaN  
1              NaN                   NaN  
2              NaN                   NaN  
3            Irish                   7.0  
4              NaN                   NaN  
5              NaN                   NaN

In [12]:
#SCARBOROUGH population distribution by ethnicity 
Scar_population_df = pd.read_html(html, header = 0)[15]
Scar_population_df = Scar_population_df.rename(columns={'%':'Ethnic Origin 1 in %', 
                                                      '%.1':'Ethnic Origin 2 in %',
                                                     '%.2':'Ethnic Origin 3 in %',
                                                     '%.3':'Ethnic Origin 4 in %',
                                                     '%.4':'Ethnic Origin 5 in %',
                                                     '%.5':'Ethnic Origin 6 in %',
                                                     '%.6':'Ethnic Origin 7 in %',
                                                     '%.7':'Ethnic Origin 8 in %'})
Scar_population_df

Riding  Population Ethnic Origin #1  Ethnic Origin 1 in %  \
0  Etobicoke-Lakeshore      127520          English                  17.1   
1      Etobicoke North      116960      East Indian                  22.2   
2     Etobicoke Centre      116055          Italian                  15.1   
3    York South-Weston      115130       Portuguese                  14.5   

  Ethnic Origin #2  Ethnic Origin 2 in % Ethnic Origin #3  \
0         Canadian                  15.9            Irish   
1         Canadian                   7.9              NaN   
2          English                  14.3         Canadian   
3          Italian                  12.8         Canadian   

   Ethnic Origin 3 in % Ethnic Origin #4  Ethnic Origin 4 in %  \
0                  14.4         Scottish                  13.5   
1                   NaN              NaN                   NaN   
2                  12.1            Irish                  10.8   
3                   8.7         Jamaican                   8.4   

  Ethnic Origin #5  Ethnic Origin 5 in % Ethnic Origin #6  \
0           Polish                   9.2          Italian   
1              NaN                   NaN              NaN   
2         Scottish                  10.4        Ukrainian   
3              NaN                   NaN              NaN   

   Ethnic Origin 6 in % Ethnic Origin #7  Ethnic Origin 7 in %  \
0                   9.1        Ukrainian                   7.6   
1                   NaN              NaN                   NaN   
2                   8.1           Polish                   7.4   
3                   NaN              NaN                   NaN   

  Ethnic Origin #8  Ethnic Origin 8 in %  
0           German                   7.1  
1              NaN                   NaN  
2              NaN                   NaN  
3              NaN                   NaN

In [13]:
#ETOBICOKE & YORK population distribution by ethnicity 
ETY_population_df = pd.read_html(html, header = 0)[16]
ETY_population_df = ETY_population_df.rename(columns={'%':'Ethnic Origin 1 in %', 
                                                      '%.1':'Ethnic Origin 2 in %',
                                                     '%.2':'Ethnic Origin 3 in %',
                                                     '%.3':'Ethnic Origin 4 in %',
                                                     '%.4':'Ethnic Origin 5 in %',
                                                     '%.5':'Ethnic Origin 6 in %',
                                                     '%.6':'Ethnic Origin 7 in %',
                                                     '%.7':'Ethnic Origin 8 in %'})
ETY_population_df



Ethnic groups in the Toronto CMA (2016)Source: Focus on Geography Series, 2016 Census; Toronto, (CMA) - Ontario  \
0                                        Ethnic group                                                                
1                                        Ethnic group                                                                
2                                        Ethnic group                                                                
3                                        Ethnic group                                                                
4                                        Ethnic group                                                                
5                                        Ethnic group                                                                
6                                        Ethnic group                                                                
7                                        Ethnic group                                                                
8                                        Ethnic group                                                                
9                                        Ethnic group                                                                
10                                       Ethnic group                                                                
11                                       Ethnic group                                                                
12                                       Ethnic group                                                                
13                                                NaN                                                                
14                                   Total population                                                                

   Ethnic groups in the Toronto CMA (2016)Source: Focus on Geography Series, 2016 Census; Toronto, (CMA) - Ontario.1  \
0                                               White                                                                  
1                                         South Asian                                                                  
2                                             Chinese                                                                  
3                                               Black                                                                  
4                                            Filipino                                                                  
5                                      Latin American                                                                  
6                                          West Asian                                                                  
7                                                Arab                                                                  
8                                     Southeast Asian                                                                  
9                                              Korean                                                                  
10                                           Japanese                                                                  
11                                Multiple minorities                                                                  
12                                              Other                                                                  
13                                         Aboriginal                                                                  
14                                   Total population                                                                  

    Population  Ethnic Origin 1 in %  
0      2804630                  47.8  
1       973225                  16.6  
2       631045                  10.8  
3       442020                   7.5 

In [36]:
TEY_population_df.head(1)


Riding  Population Ethnic Origin #1  Ethnic Origin 1 in %  \
0  Willowdale      117405          Chinese                  25.9   

  Ethnic Origin #2  Ethnic Origin 2 in % Ethnic Origin #3  \
0          Iranian                  12.1           Korean   

   Ethnic Origin 3 in % Ethnic Origin #4  Ethnic Origin 4 in %  \
0                  10.6              NaN                   NaN   

  Ethnic Origin #5  Ethnic Origin 5 in % Ethnic Origin #6  \
0              NaN                   NaN              NaN   

   Ethnic Origin 6 in % Ethnic Origin #7  Ethnic Origin 7 in %  \
0                   NaN              NaN                   NaN   

  Ethnic Origin #8  Ethnic Origin 8 in %  
0              NaN                   NaN

In [37]:
North_population_df.head(1)

Riding  Population Ethnic Origin #1  Ethnic Origin 1 in %  \
0  Scarborough Centre      110450         Filipino                  13.1   

  Ethnic Origin #2  Ethnic Origin 2 in % Ethnic Origin #3  \
0      East Indian                  12.2         Canadian   

   Ethnic Origin 3 in % Ethnic Origin #4  Ethnic Origin 4 in %  \
0                  11.2          Chinese                  10.7   

  Ethnic Origin #5  Ethnic Origin 5 in % Ethnic Origin #6  \
0          English                   7.8       Sri Lankan   

   Ethnic Origin 6 in % Ethnic Origin #7  Ethnic Origin 7 in %  \
0                   7.0              NaN                   NaN   

  Ethnic Origin #8  Ethnic Origin 8 in %  
0              NaN                   NaN

In [38]:
Scar_population_df.head(1)

Riding  Population Ethnic Origin #1  Ethnic Origin 1 in %  \
0  Etobicoke-Lakeshore      127520          English                  17.1   

  Ethnic Origin #2  Ethnic Origin 2 in % Ethnic Origin #3  \
0         Canadian                  15.9            Irish   

   Ethnic Origin 3 in % Ethnic Origin #4  Ethnic Origin 4 in %  \
0                  14.4         Scottish                  13.5   

  Ethnic Origin #5  Ethnic Origin 5 in % Ethnic Origin #6  \
0           Polish                   9.2          Italian   

   Ethnic Origin 6 in % Ethnic Origin #7  Ethnic Origin 7 in %  \
0                   9.1        Ukrainian                   7.6   

  Ethnic Origin #8  Ethnic Origin 8 in %  
0           German                   7.1

In [39]:
ETY_population_df.head(1)

Ethnic groups in the Toronto CMA (2016)Source: Focus on Geography Series, 2016 Census; Toronto, (CMA) - Ontario  \
0                                       Ethnic group                                                                

  Ethnic groups in the Toronto CMA (2016)Source: Focus on Geography Series, 2016 Census; Toronto, (CMA) - Ontario.1  \
0                                              White                                                                  

   Population  Ethnic Origin 1 in %  
0     2804630                  47.8

<h4>d) Get location data using Foursquare</h4>
Foursquare API is very usefule online application used my many developers & other application like Uber etc. In this project I have used it to retrieve informtion about the places present in the neighborhoods of Toronto. The API returns a JSON file and we need to turn that into a data-frame. Here I’ve chosen 100 popular spots for each neighborhood within a radius of 1km. 

In [14]:
#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values


In [15]:
#Use geopy library to get the latitude and longitude values of New York City.

address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [16]:
#!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library
import requests

In [17]:
CLIENT_ID = 'UTDQ4LGBHDJFK40GGRV5EZUUFFV4QBUNAAME34EGLA3ZTCWB'
CLIENT_SECRET = 'G1HQS35NYI1S5RE5TWCM0BZ10WKIC1OGMZ324LOGF3XDBOI3'
VERSION = '20200601'

In [18]:
radius=1000
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius)


In [19]:
results = requests.get(url).json()

In [20]:
#Function to get the category

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']

    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Using the get_category_type function, we clean up the json and turn it into a pandas dataframe. Before we start that we need to import certain libraries.

In [21]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']

nearby_venues = json_normalize(venues) # flatten JSON

filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name       categories        lat        lng
0          Downtown Toronto     Neighborhood  43.653232 -79.385296
1    Nathan Phillips Square            Plaza  43.652270 -79.383516
2                    Indigo        Bookstore  43.653515 -79.380696
3              Chatime 日出茶太  Bubble Tea Shop  43.655542 -79.384684
4  Textile Museum of Canada       Art Museum  43.654396 -79.386500

Now we can explore the nearby venues!

In [22]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [23]:
LIMIT = 100
toronto_venues = getNearbyVenues(names=toronto_DF['Neighborhood'],
                                   latitudes=toronto_DF['Latitude'],
                                   longitudes=toronto_DF['Longitude']
                                  )

Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government
The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Business reply mail Processing Centre, South Central Letter 

In [28]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Lawrence Park              43.728020              -79.388790   
1     Lawrence Park              43.728020              -79.388790   
2     Lawrence Park              43.728020              -79.388790   
3  Davisville North              43.712751              -79.390197   
4  Davisville North              43.712751              -79.390197   

                            Venue  Venue Latitude  Venue Longitude  \
0            Lawrence Park Ravine       43.726963       -79.394382   
1              Zodiac Swim School       43.728532       -79.382860   
2  TTC Bus #162 - Lawrence-Donway       43.728026       -79.382805   
3     Homeway Restaurant & Brunch       43.712641       -79.391557   
4                   Sherwood Park       43.716551       -79.387776   

   Venue Category  
0            Park  
1     Swim School  
2        Bus Line  
3  Breakfast Spot  
4            Park

In [29]:
toronto_venues.groupby('Neighborhood').count()


Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     20   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   24   
...                                                                ...   
Willowdale, Willowdale West                                          5   
Woburn                                                               5   
Woodbine Heights                                                     9   
York Mills West                                                      3   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      20   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    24   
...                                                                 ...   
Willowdale, Willowdale West                                           5   
Woburn                                                                5   
Woodbine Heights                                                      9   
York Mills West                                                       3   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     20              20   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   24              24   
...                                                ...             ...   
Willowdale, Willowdale West                          5               5   
Woburn                                               5               5   
Woodbine Heights                                     9               9   
York Mills West                                      3               3   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               20   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             24   
...                                                          ...   
Willowdale, Willowdale West                                    5   
Woburn                                                         5   
Woodbine Heights                                               9   
York Mills West                                                3   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

In [30]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 272 uniques categories.


There are 272 unique categories in which Indian Restaurant is one of them. We will do one hot encoding for getting dummies of venue category. So that we will calculate mean of all venue groupby there neighborhoods.


In [31]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped



Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
91                      Willowdale, Willowdale West          0.0   
92                                           Woburn          0.0   
93                                 Woodbine Heights          0.0   
94                                  York Mills West          0.0   
95                         York Mills, Silver Hills          0.0   

    Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                 0.0                0.0      0.0                 0.0   
1                 0.0                0.0      0.0                 0.0   
2                 0.0                0.0      0.0                 0.0   
3                 0.0                0.0      0.0                 0.0   
4                 0.0                0.0      0.0                 0.0   
..                ...                ...      ...                 ...   
91                0.0                0.0      0.0                 0.0   
92                0.0                0.0      0.0                 0.0   
93                0.0                0.0      0.0                 0.0   
94                0.0                0.0      0.0                 0.0   
95                0.0                0.0      0.0                 0.0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0            0.0             0.0              0.0               0.0  ...   
1            0.0             0.0              0.0               0.0  ...   
2            0.0             0.0              0.0               0.0  ...   
3            0.0             0.0              0.0               0.0  ...   
4            0.0             0.0              0.0               0.0  ...   
..           ...             ...              ...               ...  ...   
91           0.0             0.0              0.0               0.0  ...   
92           0.0             0.0              0.0               0.0  ...   
93           0.0             0.0              0.0               0.0  ...   
94           0.0             0.0              0.0               0.0  ...   
95           0.0             0.0              0.0               0.0  ...   

    Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0     0.0            0.0                            0.0               0.0   
1     0.0            0.0                            0.0               0.0   
2     0.0            0.0                            0.0               0.0   
3     0.0            0.0                            0.0               0.0   
4     0.0            0.0                            0.0               0.0   
..    ...            ...                            ...               ...   
91    0.0            0.0                            0.0               0.0   
92    0.0            0.0                            0.0               0.0   
93    0.0            0.0                            0.0               0.0   
94    0.0            0.0                            0.0               0.0   
95    0.0            0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000                    0.0              0.0       0.0   
1      0.000000                    0.0              0.0       0.0   
2      0.000000                    0.0              0.0       0.0   
3      0.000000                    0.0              0.0       0.0   
4      0.000000                    0.0              0.0       0.0   
..          ...                    ...              ...       ...   
91     0.000000         

In [34]:
print (toronto_venues['Venue Category'].value_counts())


Coffee Shop        177
Café                99
Restaurant          68
Pizza Place         52
Park                50
                  ... 
Drugstore            1
Skate Park           1
Hostel               1
Harbor / Marina      1
Bed & Breakfast      1
Name: Venue Category, Length: 272, dtype: int64
